source/grid/tria_description.cc

```diff
@@ -0,0 +1,662 @@
+        const typename CellId::binary_type &binary_representation)
+      {
+        // exploiting the structure of CellId::binary_type
+        // see also the documentation of CellId
+
+        // actual coarse-grid id
+        const unsigned int coarse_cell_id  = binary_representation[0];
+        const unsigned int n_child_indices = binary_representation[1] >> 2;
+
+        const unsigned int children_per_value =
+          sizeof(CellId::binary_type::value_type) * 8 / dim;
+        unsigned int child_level  = 0;
+        unsigned int binary_entry = 2;
+
+        // path to the get to the cell
+        std::vector<unsigned int> cell_indices;
+        while (child_level < n_child_indices)
+          {
+            Assert(binary_entry < binary_representation.size(),
+                   ExcInternalError());
+
+            for (unsigned int j = 0; j < children_per_value; ++j)
+              {
+                unsigned int cell_index =
+                  (((binary_representation[binary_entry] >> (j * dim))) &
+                   (GeometryInfo<dim>::max_children_per_cell - 1));
+                cell_indices.push_back(cell_index);
+                ++child_level;
+                if (child_level == n_child_indices)
+                  break;
+              }
+            ++binary_entry;
+          }
+
+        // compute new coarse-grid id: c_{i+1} = c_{i}*2^dim + q;
+        types::coarse_cell_id level_coarse_cell_id = coarse_cell_id;
+        for (auto i : cell_indices)
+          level_coarse_cell_id =
+            level_coarse_cell_id * GeometryInfo<dim>::max_children_per_cell + i;
+
+        return level_coarse_cell_id;
+      }
+    } // namespace
+
+
+    template <int dim, int spacedim>
+    Description<dim, spacedim>
+    create_description_from_triangulation(
+      const dealii::Triangulation<dim, spacedim> &tria,
+      const MPI_Comm                              comm,
+      const bool         construct_multilevel_hierarchy,
+      const unsigned int my_rank_in)
+    {
+      if (auto tria_pdt = dynamic_cast<
+            const parallel::distributed::Triangulation<dim, spacedim> *>(&tria))
+        AssertThrow(comm == tria_pdt->get_communicator(),
+                    ExcMessage("MPI communicators do not match."));
+
+      // First, figure out for what rank we are supposed to build the
+      // ConstructionData object
+      unsigned int my_rank = my_rank_in;
+      AssertThrow(my_rank == numbers::invalid_unsigned_int ||
+                    my_rank < dealii::Utilities::MPI::n_mpi_processes(comm),
+                  ExcMessage(
+                    "Rank has to be smaller than available processes."));
+
+      if (auto tria_pdt = dynamic_cast<
+            const parallel::distributed::Triangulation<dim, spacedim> *>(&tria))
+        {
+          AssertThrow(
+            my_rank == numbers::invalid_unsigned_int ||
+              my_rank == dealii::Utilities::MPI::this_mpi_process(comm),
```

Comment:
Shouldn't this be true also for a `shared` triangulation?

Reply:
I have never tried out shared triangulations. But I guess this assert does not apply for it. My guess is that the next branch is taken and it should work fine if it is treated like a serial triangulation.